File: py/src/instructions/group_create.py
from __future__ import annotations
import typing
from solana.publickey import PublicKey
from solana.transaction import TransactionInstruction, AccountMeta
import borsh_construct as borsh
from ..program_id import PROGRAM_ID


class GroupCreateArgs(typing.TypedDict):
    group_num: int
    testing: int
    version: int


layout = borsh.CStruct(
    "group_num" / borsh.U32, "testing" / borsh.U8, "version" / borsh.U8
)


class GroupCreateAccounts(typing.TypedDict):
    group: PublicKey
    creator: PublicKey
    insurance_mint: PublicKey
    insurance_vault: PublicKey
    payer: PublicKey
    token_program: PublicKey
    system_program: PublicKey
    rent: PublicKey


def group_create(
    args: GroupCreateArgs,
    accounts: GroupCreateAccounts,
    program_id: PublicKey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> TransactionInstruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["group"], is_signer=False, is_writable=True),
        AccountMeta(pubkey=accounts["creator"], is_signer=True, is_writable=False),
        AccountMeta(
            pubkey=accounts["insurance_mint"], is_signer=False, is_writable=False
        ),
        AccountMeta(
            pubkey=accounts["insurance_vault"], is_signer=False, is_writable=True
        ),
        AccountMeta(pubkey=accounts["payer"], is_signer=True, is_writable=True),
        AccountMeta(
            pubkey=accounts["token_program"], is_signer=False, is_writable=False
        ),
        AccountMeta(
            pubkey=accounts["system_program"], is_signer=False, is_writable=False
        ),
        AccountMeta(pubkey=accounts["rent"], is_signer=False, is_writable=False),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b"\xe2R\xefwk\x88\xa6\xf0"
    encoded_args = layout.build(
        {
            "group_num": args["group_num"],
            "testing": args["testing"],
            "version": args["version"],
        }
    )
    data = identifier + encoded_args
    return TransactionInstruction(keys, program_id, data)

File: programs/mango-v4/src/instructions/group_create.rs
use anchor_lang::prelude::*;

use crate::accounts_ix::*;
use crate::error::*;

pub fn group_create(
    ctx: Context<GroupCreate>,
    group_num: u32,
    testing: u8,
    version: u8,
) -> Result<()> {
    let mut group = ctx.accounts.group.load_init()?;
    group.creator = ctx.accounts.creator.key();
    group.group_num = group_num;
    group.admin = ctx.accounts.creator.key();
    group.fast_listing_admin = Pubkey::default();
    group.insurance_vault = ctx.accounts.insurance_vault.key();
    group.insurance_mint = ctx.accounts.insurance_mint.key();
    group.bump = *ctx.bumps.get("group").ok_or(MangoError::SomeError)?;
    group.testing = testing;
    group.version = version;
    Ok(())
}
